src/types/Course/Preference.ts

```diff
@@ -35,38 +72,218 @@ export const PreferenceQuery = extendType({
+              id: {
+                subject: courseData.subject,
+                code: courseData.code,
+                term: courseData.term,
+                year: courseData.year,
+              },
+              preference: rank ?? 0,
+            };
+          })
+        );
+        return courses;
+      },
+    });
+  },
+});
+
+export const PreferenceMutation = extendType({
+  type: 'Mutation',
+  definition(t) {
+    t.nonNull.field('createTeachingPreference', {
+      type: Response,
+      description: 'Teaching preferences',
+      args: {
+        input: arg({ type: nonNull(CreateTeachingPreferenceInput) }),
+      },
+      resolve: async (
+        _,
+        {
+          input: {
+            peng,
+            userId,
+            year,
+            courses,
+            hasRelief,
+            reliefReason,
+            hasTopic,
+            topicDescription,
+            maxFallCapacity,
+            maxSpringCapacity,
+            maxSummerCapacity,
+          },
+        },
+        { prisma }
+      ) => {
+        const preferenceObjs = await Promise.all(
+          courses.map(
+            async ({
+              subject,
+              code,
+              term,
+              preference,
+            }: {
+              subject: string;
+              code: string;
+              term: Term;
+              preference: number;
+            }) => {
+              const courseData = await (prisma as PrismaClient).course.findFirst({
+                where: {
+                  subject: subject,
+                  code: code,
+                  term: term,
+                  year: year,
+                },
+              });
+
+              const courseID = courseData ? courseData.id : -1; //if courseData is null set id to -1
+
+              return {
+                courseID: courseID,
+                rank: preference ?? 0,
+              };
+            }
+          )
+        );
+
+        for (const preferenceObj of preferenceObjs) {
+          await (prisma as PrismaClient).preference.upsert({
+            where: {
+              // eslint-disable-next-line camelcase
+              userID_courseID: {
+                userID: Number(userId),
+                courseID: preferenceObj.courseID,
+              },
+            },
+            update: {
+              rank: preferenceObj.rank,
+            },
+            create: {
+              userID: Number(userId),
+              courseID: preferenceObj.courseID,
+              rank: preferenceObj.rank,
+            },
+          });
+        }
+
+        const prefs = await (prisma as PrismaClient).preference.findMany({
+          where: {
+            user: {
+              id: Number(userId),
+            },
+          },
+        });
+
+        await (prisma as PrismaClient).user.update({
+          where: {
+            id: Number(userId),
+          },
+          data: {
+            peng: peng,
+            preferences: {
+              set: prefs,
```

Comment:
```suggestion
              set: prefs.map(({userID, courseID}) => ({userID_courseID: {userID, courseID}}))
```
this should work. It's because the `set` attribute takes in as an object the unique identifiers of what is being set. So for preferences, the unique identifiers are userID and courseID. The `Preference` type you were using before has `rank` associated with it as well which isn't unique.